File: tests/test_drive_links.js
'use strict';
const test=require('node:test'),assert=require('node:assert/strict');
const links=require('../static/drive-links.js');
const folder='https://drive.google.com/drive/u/0/folders/folder-one?usp=sharing&resourcekey=key-one';
const file='https://drive.google.com/file/d/file-one/view?usp=sharing';
test('recognizes shared Drive resources and preserves access keys and sheet tabs',()=>{
 assert.equal(links.parse(folder).type,'文件夹');assert.match(links.parse(folder).url,/resourcekey=key-one/);
 assert.equal(links.parse(file).key,links.parse('https://drive.google.com/open?id=file-one').key);
 const sheet='https://docs.google.com/spreadsheets/d/sheet-one/edit#gid=42';
 assert.equal(links.parse(sheet).type,'表格');assert.equal(links.parse(sheet).url,sheet);
 for(const bad of ['javascript:alert(1)','http://drive.google.com/file/d/id/view','https://drive.google.com.evil.test/file/d/id/view','https://drive.google.com@evil.test/file/d/id/view','https://evil@drive.google.com/file/d/id/view','https://drive.google.com:444/file/d/id/view','https://drive.google.com\\@evil.test/file/d/id/view','https://drive.google.com/','https://docs.google.com/redirect?url=https://evil.test'])assert.equal(links.parse(bad),null,bad);
});
test('extracts old pasted links without turning arbitrary URLs or text into markup',()=>{
 const body=`大型资料\n${folder}。\n(${file})\n${file}\n<script>alert(1)</script> https://evil.test/`;
 assert.equal(links.extract(body).length,2);
 assert.equal(links.bodyText(body).includes(folder),false);
 assert.match(links.bodyText(body),/<script>alert\(1\)<\/script> https:\/\/evil.test/);
 assert.equal(links.bodyText('纯文字'), '纯文字');
});
test('aggregates only this project, deduplicates pasted references, and retains section sources',()=>{
 const records=[
  {id:'stored',kind:'files',project_id:'p',title:'Known file',url:file},
  {id:'upload',kind:'files',project_id:'p',storage_provider:'gcs',url:'/api/files/upload/content'},
  {id:'n1',kind:'notes',project_id:'p',subsection_id:'s1',body:`${folder}\n${file}`,attachment_ids:['stored','upload']},
  {id:'n2',kind:'notes',project_id:'p',subsection_id:'s2',body:folder,attachment_ids:[]},
  {id:'secret',kind:'notes',project_id:'other',body:'https://drive.google.com/file/d/secret/view'},
 ];
 const before=JSON.stringify(records),files=links.projectFiles(records,'p');
 assert.equal(files.length,3);assert.equal(JSON.stringify(records),before);
 assert.equal(files.filter(f=>f.virtual).length,1);
 assert.deepEqual(files.find(f=>f.virtual).source_note_ids,['n1','n2']);
 assert.equal(links.noteFiles(records,records[2]).length,3);
 assert.equal(links.noteFiles(records,records[3]).length,1);
 const removed=records.map(r=>r.kind==='notes'?{...r,body:'no links'}:r);
 assert.equal(links.projectFiles(removed,'p').length,2);
});

File: tests/test_drive_links.py
"""Drive link creation and optional note attachment are one scoped transaction."""
import uuid
import pytest
from test_app import client, ready, login
from test_project_deletion import deletion_client, delete
from test_resource_access import created
from test_access import account, as_admin, as_member, grant

@pytest.fixture
def data(deletion_client):
    c=deletion_client;ready(c)
    p=created(c,'projects')
    n=created(c,'notes',project_id=p['id'],body='Synthetic note',date='2026-09-11')
    return c,p,n

def link(p,n=None,**extra):
    return {'title':'Synthetic Drive folder','url':'https://drive.google.com/drive/folders/synthetic-folder?resourcekey=synthetic-key','project_id':p['id'],'note_id':n['id'] if n else '', 'creation_key':str(uuid.uuid4()),**extra}

def state(c):return {r['id']:r for r in c.get('/api/state').json()['records']}

def test_link_retry_audits_and_intentional_unlink(data):
    c,p,n=data;body=link(p,n)
    first=c.post('/api/records/files',json=body);assert first.status_code==200
    f=first.json();assert state(c)[n['id']]['attachment_ids']==[f['id']]
    assert c.post('/api/records/files',json=body).json()['id']==f['id']
    assert state(c)[n['id']]['version']==2
    audit=c.get('/api/activity?entity_id='+n['id']).json()[0]
    assert audit['actor_name']=='Kevin' and audit['action']=='关联附件'
    assert audit['before']['attachment_ids']==[] and audit['after']['attachment_ids']==[f['id']]
    assert c.patch('/api/records/notes/'+n['id'],json={'version':2,'attachment_ids':[]}).status_code==200
    assert c.post('/api/records/files',json=body).status_code==409
    assert state(c)[n['id']]['attachment_ids']==[]
    assert delete(c,p).status_code==200
    assert c.post('/api/records/files',json=body).status_code in (403,409)

def test_link_cannot_cross_scopes_or_escalate_viewer(data):
    c,p,n=data;other=created(c,'projects')
    private=created(c,'notes',project_id=other['id'],body='Private note',date='2026-09-11')
    member=account(c);grant(c,member,'projects',p['id'],'viewer');as_member(c)
    assert c.post('/api/records/files',json=link(p,n)).status_code==403
    as_admin(c);grant(c,member,'projects',p['id'],'editor',1);assert login(c,'chloe','changed-member-123!').status_code==200
    assert c.post('/api/records/files',json=link(p,private)).status_code==404
    assert c.post('/api/records/files',json=link(p,n)).status_code==200
    assert c.post('/api/records/files',json=link(other)).status_code==403
    as_admin(c)
    assert c.post('/api/records/files',json=link(p,private)).status_code==409
    c.headers['X-Cmp-Workspace']='personal:'+member['id']
    assert c.post('/api/records/files',json=link(p,n)).status_code==403

def test_link_audit_failure_rolls_back_file_and_note(data,monkeypatch):
    import app
    c,p,n=data;original=app.audit
    def broken(c,actor,action,kind,*args,**kwargs):
        if kind=='notes':raise RuntimeError('Synthetic audit failure')
        return original(c,actor,action,kind,*args,**kwargs)
    monkeypatch.setattr(app,'audit',broken)
    with pytest.raises(RuntimeError):c.post('/api/records/files',json=link(p,n))
    assert not [r for r in state(c).values() if r['kind']=='files']
    assert state(c)[n['id']]['version']==1
    assert not [a for a in c.get('/api/activity').json() if a['kind']=='files']

def test_link_rejects_unsafe_urls_and_stale_section(data):
    c,p,n=data
    for url in ['javascript:alert(1)','https://drive.google.com.evil.test/file/d/id/view','https://user:pass@drive.google.com/file/d/id/view','https://drive.google.com:bad/file/d/id/view','https://drive.google.com\\@evil.test/file/d/id/view']:
        assert c.post('/api/records/files',json=link(p,url=url)).status_code==400
    section=created(c,'subsections',project_id=p['id'])
    assert c.patch('/api/records/notes/'+n['id'],json={'version':1,'subsection_id':section['id']}).status_code==200
    assert c.post('/api/records/files',json=link(p,n)).status_code==409
    assert not [r for r in state(c).values() if r['kind']=='files']
